plugins/modules/system/osx_defaults.py

```diff
@@ -261,49 +294,37 @@ def read(self):
         # Ok, lets parse the type from output
         data_type = out.strip().replace('Type is ', '')
 
-        # Now get the current value
-        rc, out, err = self.module.run_command(self._base_command() + ["read", self.domain, self.key])
-
-        # Strip output
-        out = out.strip()
+        # Export plist in xml
+        rc, out, err = self.module.run_command(self._base_command() + ["export", self.domain, '-'])
 
-        # An non zero RC at this point is kinda strange...
         if rc != 0:
-            raise OSXDefaultsException("An error occurred while reading key value from defaults: %s" % out)
+            raise OSXDefaultsException("An error occurred while exporting plist: %s" % out)
 
-        # Convert string to list when type is array
-        if data_type == "array":
-            out = self._convert_defaults_str_to_list(out)
+        # Convert xml to Python objects
+        if PY2:
+            self.current_value = plistlib.readPlistFromString(out)
+        else:
+            self.current_value = plistlib.loads(out.encode('utf-8'))
```

Comment:
It might be a good idea to move this into a static function (that can easily be tested), and add unit tests for this. These are run with all Python versions Ansible supports, which will give you some confidence that this does what it is supposed to on all Python versions. Unit tests for this module would have to do into tests/unit/plugins/modules/system/test_osx_defaults.py (which doesn't exist yet).

Reply:
I'm not very experienced with unit tests in Python. I read some of tests/unit/plugins/modules/system/* but it seems difficult to understand. Is there good document for this?

Reply:
> move this into a static function

Like this?
```py
def convert_plist_xml_to_object(xml):
    if six.PY2:
        return plistlib.readPlistFromString(xml)
    else:
        return plistlib.loads(xml.encode('utf-8'))
```

Reply:
You can find some information here: https://docs.ansible.com/ansible/latest/dev_guide/testing_units.html#structuring-unit-tests

And yes, such a function would help.